Refactor ForeachUnaryOp.cu

izdeby opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #49039 [wip] Replace optimizers in torch.optim with the ones from torch.optim._multi_tensor
* #48224 [wip] Pushed unsupported scenarios to slow path
* #51170 Updated foreach pointwise ops to support complex and bloat16 on CUDA
* #49714 Update foreach max/min
* #51060 Refactor foreach pointwise ops tests to use OpInfo
* #51059 Refactor foreach binary ops tests with tensor lists to use OpInfo
* #51058 Refactor foreach binary ops tests with scalars to use OpInfo
* #49712 Refactor foreach unary ops tests to use OpInfo
* #49250 Update foreach_div and foreach_sub logic
* #49249 Update foreach binary ops with a scalar list
* #51131 Refactor ForeachUtils.h
* **#49248 Refactor ForeachUnaryOp.cu**
* #48223 Update foreach APIs to use scalar lists

Differential Revision: [D25502940](https://our.internmc.facebook.com/intern/diff/D25502940)

----------
- Refactor ForeachUnaryOps.cu to better see what dtypes are supported for ops.

Commit:
Refactor ForeachUnaryOps.cu

[ghstack-poisoned]

Commit:
Update on "Refactor ForeachUnaryOps.cu"


Differential Revision: [D25502940](https://our.internmc.facebook.com/intern/diff/D25502940)

----------
- Refactor ForeachUnaryOps.cu to better see what dtypes are supported for ops.
- Add `has_bool_tensor` checks that will be crucial in next PRs in this stack

[ghstack-poisoned]

Commit:
Update on "Refactor ForeachUnaryOps.cu"


Differential Revision: [D25502940](https://our.internmc.facebook.com/intern/diff/D25502940)

----------
- Refactor ForeachUnaryOps.cu to better see what dtypes are supported for ops.
- Add `has_bool_tensor` checks that will be crucial in next PRs in this stack

[ghstack-poisoned]

Commit:
Update on "Refactor ForeachUnaryOps.cu"


Differential Revision: [D25502940](https://our.internmc.facebook.com/intern/diff/D25502940)

----------
- Refactor ForeachUnaryOps.cu to better see what dtypes are supported for ops.
- Add `has_bool_tensor` checks that will be crucial in next PRs in this stack

[ghstack-poisoned]

Commit:
Update on "Refactor ForeachUnaryOps.cu"


Differential Revision: [D25502940](https://our.internmc.facebook.com/intern/diff/D25502940)

----------
- Refactor ForeachUnaryOps.cu to better see what dtypes are supported for ops.
- Add `has_bool_tensor` checks that will be crucial in next PRs in this stack

[ghstack-poisoned]

Commit:
Update on "Refactor ForeachUnaryOp.cu"


Differential Revision: [D25502940](https://our.internmc.facebook.com/intern/diff/D25502940)

----------
- Refactor ForeachUnaryOps.cu to better see what dtypes are supported for ops.
- Add `has_bool_tensor` checks that will be crucial in next PRs in this stack

[ghstack-poisoned]

Commit:
Update on "Refactor ForeachUnaryOp.cu"


Differential Revision: [D25502940](https://our.internmc.facebook.com/intern/diff/D25502940)

----------
- Refactor ForeachUnaryOps.cu to better see what dtypes are supported for ops.

[ghstack-poisoned]

zou3519 commented on aten/src/ATen/native/cuda/ForeachUnaryOp.cu:
```diff
@@ -85,7 +85,7 @@ void foreach_unary_op_complex_bfloat16_(TensorList tensors) {
 
 template <template<class> class Op>
-std::vector<Tensor> foreach_unary_op(TensorList tensors) {
+std::vector<Tensor> all_types_half_bfloat16(TensorList tensors) {
```
This is new, right? Does this PR add additional behavior to any of the foreach ops? If so we should note it in the PR description and add tests for those behaviors

zou3519 commented on aten/src/ATen/native/cuda/ForeachUnaryOp.cu:
```diff
@@ -180,150 +204,164 @@ void foreach_unary_op_bfloat16_(TensorList tensors) {
+    check_foreach_api_restrictions(tensors);                             \
     if (!can_use_fast_route(tensors)) {                                 \
-        return at::native::foreach_tensor_##NAME##_slow(tensors);       \
+        at::native::foreach_tensor_##NAME##_slow(tensors);              \
```
Why the deletion of the `return`?

zou3519 commented on aten/src/ATen/native/cuda/ForeachUnaryOp.cu:
```diff
@@ -180,150 +204,164 @@ void foreach_unary_op_bfloat16_(TensorList tensors) {
+};
+FLOATING_HALF_BFLOAT16(log1p, Log1p);
+FLOATING_HALF_BFLOAT16(erf, Erf);
+FLOATING_HALF_BFLOAT16(sigmoid, Sigmoid);
```
sigmoid is new, right?

zou3519 replied:
Nvm, I see it now

zou3519 commented on aten/src/ATen/native/cuda/ForeachUnaryOp.cu:
```diff
@@ -180,150 +204,164 @@ void foreach_unary_op_bfloat16_(TensorList tensors) {
-    }                                                                   \
-                                                                        \
-    foreach_unary_op_bfloat16_<NAME1>(tensors);                         \
+#define FLOATING_HALF(NAME, NAME1)                                                 \
+template<typename T>                                                               \
+struct NAME1 {                                                                     \
+    __device__ T operator()(T t) const { return std::NAME(t); }                    \
+};                                                                                 \
+                                                                                   \
+std::vector<Tensor> foreach_tensor_##NAME##_cuda(TensorList tensors) {             \
+    check_foreach_api_restrictions(tensors);                                       \
+    if (!can_use_fast_route(tensors)) {                                            \
+        return at::native::foreach_tensor_##NAME##_slow(tensors);                  \
+    }                                                                              \
+                                                                                   \
+                                                                                   \
+    return floating_half<NAME1>(tensors);                                          \
+}                                                                                  \
+                                                                                   \
+void foreach_tensor_##NAME##_cuda_(TensorList tensors) {                           \
+    check_foreach_api_restrictions(tensors);                                       \
+    if (!can_use_fast_route(tensors)) {                                            \
+        at::native::foreach_tensor_##NAME##_slow_(tensors);                        \
+    }                                                                              \
+                                                                                   \
+    floating_half_<NAME1>(tensors);                                                \
```
It seems to me what you really want here is three separate macros:
```
// makes the functor
#define STD_FUNCTOR(op_name, functor_name)
template<typename T>
struct NAME1 {
    __device__ T operator()(T t) const { return std::NAME(t); }
};

// given a functor and a "dispatch function", creates the outplace and inplace operations
#define OP_CUSTOM_FUNCTOR(function, op_name, functor_name)
std::vector<Tensor> foreach_tensor_##op_name##_cuda(TensorList tensors) {
    check_foreach_api_restrictions(tensors);
    if (!can_use_fast_route(tensors)) {
        return at::native::foreach_tensor_##op_name##_slow(tensors);
    }
    return function<functor_name>(tensors);
}
void foreach_tensor_##op_name##_cuda_(TensorList tensors) {
    check_foreach_api_restrictions(tensors);
    if (!can_use_fast_route(tensors)) {
        return at::native::foreach_tensor_##op_name##_slow_(tensors);
    }

    ##function##_<functor_name>(tensors);
}

// creates a functor, outplace version, and inplace version.
#define OP(function, op_name, functor_name)
STD_FUNCTOR(op_name, functor_name);
OP_CUSTOM_FUNCTOR(function, op_name, functor_name);
```

And then you can write code like:
```
OP(floating_half, erfc, Erfc);
OP(floating_half, expm1, Expm1);
OP(floating_half_bfloat16, log1p, Log1p);
...
```

zou3519 commented on aten/src/ATen/native/cuda/ForeachUnaryOp.cu:
```diff
@@ -180,150 +204,164 @@ void foreach_unary_op_bfloat16_(TensorList tensors) {
 // Special cases
 //
 std::vector<Tensor> foreach_tensor_neg_cuda(TensorList tensors) {
     check_foreach_api_restrictions(tensors);
+    TORCH_CHECK(tensors[0].scalar_type() != kBool,
+              "Negation, the `-` operator, on a bool tensor is not supported. "
+              "If you are trying to invert a mask, use the `~` or `logical_not()` operator instead.");
 
     if (!can_use_fast_route(tensors)) {
         return at::native::foreach_tensor_neg_slow(tensors);
     }
 
-    return foreach_unary_op_complex_bfloat16<std::negate>(tensors);
+    return all_types_half_bfloat16<std::negate>(tensors);
```
Why is the new check needed? Also, this updates the semantics of neg, right?

zou3519 commented on aten/src/ATen/native/cuda/ForeachUnaryOp.cu:
```diff
@@ -348,47 +386,49 @@ void foreach_tensor_round_cuda_(TensorList tensors) {
     for (auto t : tensors) {
-        if (at::isComplexType(t.scalar_type())) {
-            has_complex = true;
+        if (at::isComplexType(t.scalar_type()) || 
+            at::isIntegralType(t.scalar_type(), /*includeBool=*/true)) {
+            has_complex_or_integer = true;
```
This is fixing a bug, right? Should we add a test for it?

zou3519 commented on aten/src/ATen/native/cuda/ForeachUnaryOp.cu:
```diff
@@ -180,150 +204,164 @@ void foreach_unary_op_bfloat16_(TensorList tensors) {
+FLOATING_HALF(floor, Floor);
+FLOATING_HALF(ceil, Ceil);
+
+template<typename T>
+struct Log1p {
+    __device__ T operator()(T t) const { return std::log1p(t); }
+};
+
+template<typename T>
+struct Erf {
+    __device__ T operator()(T t) const { return std::erf(t); }
+};
```
We didn't have to write these out before, but now we do :(

zou3519 replied:
The suggestion here https://github.com/pytorch/pytorch/pull/49248#discussion_r545216645 offers a way to fix this; in particular, it creates two macros: one with a way to register a custom (non-std-based) Functor, and one that just autogenerates a std:: based functor.

zou3519 reviewed (commented):
Have some questions and suggestions

Commit:
Update on "Refactor ForeachUnaryOp.cu"


Differential Revision: [D25502940](https://our.internmc.facebook.com/intern/diff/D25502940)

----------
- Refactor ForeachUnaryOps.cu to better see what dtypes are supported for ops.

[ghstack-poisoned]

zou3519 commented on aten/src/ATen/native/cuda/ForeachUnaryOp.cu:
```diff
@@ -16,304 +15,189 @@ std::vector<Tensor> foreach_unary_op_complex(TensorList tensors) {
-FOREACH_UNARY_OP_COMPLEX(tan, Tan);
-FOREACH_UNARY_OP_COMPLEX(sin, Sin);
-FOREACH_UNARY_OP_COMPLEX(sinh, Sinh);
-
-FOREACH_UNARY_OP_COMPLEX_BFLOAT16(exp, Exp);
-FOREACH_UNARY_OP_COMPLEX_BFLOAT16(sqrt, Sqrt);
-FOREACH_UNARY_OP_COMPLEX_BFLOAT16(cos, Cos);
-FOREACH_UNARY_OP_COMPLEX_BFLOAT16(tanh, Tanh);
-FOREACH_UNARY_OP_COMPLEX_BFLOAT16(log, Log);
-FOREACH_UNARY_OP_COMPLEX_BFLOAT16(log10, Log10);
-FOREACH_UNARY_OP_COMPLEX_BFLOAT16(log2, Log2);
+void foreach_tensor_sigmoid_cuda_(TensorList tensors) {
+    check_foreach_api_restrictions(tensors);
+    if (!can_use_fast_route(tensors)) {
+        at::native::foreach_tensor_sigmoid_slow_(tensors);
+    }
+
+    floating_half_bfloat16_<Sigmoid>(tensors);
+}
```
Can this just be written as:
`OP_CUSTOM_FUNCTOR(floating_half_bfloat16, sigmoid, Sigmoid)` ?

izdeby replied:
Updated

zou3519 commented on aten/src/ATen/native/cuda/ForeachUnaryOp.cu:
```diff
@@ -16,304 +15,189 @@ std::vector<Tensor> foreach_unary_op_complex(TensorList tensors) {
     }
 
-    return foreach_unary_op_complex_bfloat16<std::negate>(tensors);
+    return all_types_half_complex_bfloat16<std::negate>(tensors);
```
I assume you're going to change `neg`, `round`, `trunc`, and `abs` later and that is the reason why they're not using the `OP` or `OP_CUSTOM_FUNCTOR` macros?

zou3519 reviewed (approved):
Thanks, everything is clearer now!

Commit:
Update on "Refactor ForeachUnaryOp.cu"


Differential Revision: [D25502940](https://our.internmc.facebook.com/intern/diff/D25502940)

----------
- Refactor ForeachUnaryOps.cu to better see what dtypes are supported for ops.

[ghstack-poisoned]

Commit:
Update on "Refactor ForeachUnaryOp.cu"


Differential Revision: [D25502940](https://our.internmc.facebook.com/intern/diff/D25502940)

----------
- Refactor ForeachUnaryOps.cu to better see what dtypes are supported for ops.

[ghstack-poisoned]

Commit:
Update on "Refactor ForeachUnaryOp.cu"


Differential Revision: [D25502940](https://our.internmc.facebook.com/intern/diff/D25502940)

----------
- Refactor ForeachUnaryOps.cu to better see what dtypes are supported for ops.

[ghstack-poisoned]

Commit:
Update on "Refactor ForeachUnaryOp.cu"


Differential Revision: [D25502940](https://our.internmc.facebook.com/intern/diff/D25502940)

----------
- Refactor ForeachUnaryOps.cu to better see what dtypes are supported for ops.

[ghstack-poisoned]

Commit:
Update on "Refactor ForeachUnaryOp.cu"


Differential Revision: [D25502940](https://our.internmc.facebook.com/intern/diff/D25502940)

----------
- Refactor ForeachUnaryOps.cu to better see what dtypes are supported for ops.

[ghstack-poisoned]

ngimel reviewed (commented):
This is a nice refactor! 
There is a lot of changes currently with function getting bfloat16 support, or unary functions getting support for integer outputs (which we have to disallow for for_each). 
Not for this PR, but to make detection of those changes/breakages automatic maybe we should move testing to be OpInfo based (as I noted on another PR). Let's talk about that with @mruberry.

izdeby committed:
Update on "Refactor ForeachUnaryOp.cu"


Differential Revision: [D25502940](https://our.internmc.facebook.com/intern/diff/D25502940)

----------
- Refactor ForeachUnaryOps.cu to better see what dtypes are supported for ops.

[ghstack-poisoned]

Commit:
Update on "Refactor ForeachUnaryOp.cu"


Differential Revision: [D25502940](https://our.internmc.facebook.com/intern/diff/D25502940)

----------
- Refactor ForeachUnaryOps.cu to better see what dtypes are supported for ops.

[ghstack-poisoned]

Commit:
Update on "Refactor ForeachUnaryOp.cu"


Differential Revision: [D25502940](https://our.internmc.facebook.com/intern/diff/D25502940)

----------
- Refactor ForeachUnaryOps.cu to better see what dtypes are supported for ops.

[ghstack-poisoned]

Commit:
Update on "Refactor ForeachUnaryOp.cu"


Differential Revision: [D25502940](https://our.internmc.facebook.com/intern/diff/D25502940)

----------
- Refactor ForeachUnaryOps.cu to better see what dtypes are supported for ops.

[ghstack-poisoned]

Commit:
Update on "Refactor ForeachUnaryOp.cu"


Differential Revision: [D25502940](https://our.internmc.facebook.com/intern/diff/D25502940)

----------
- Refactor ForeachUnaryOps.cu to better see what dtypes are supported for ops.

[ghstack-poisoned]

Commit:
Update on "Refactor ForeachUnaryOp.cu"


Differential Revision: [D25502940](https://our.internmc.facebook.com/intern/diff/D25502940)

----------
- Refactor ForeachUnaryOps.cu to better see what dtypes are supported for ops.

[ghstack-poisoned]

Commit:
Update on "Refactor ForeachUnaryOp.cu"


Differential Revision: [D25502940](https://our.internmc.facebook.com/intern/diff/D25502940)

----------
- Refactor ForeachUnaryOps.cu to better see what dtypes are supported for ops.

[ghstack-poisoned]

Commit:
Update on "Refactor ForeachUnaryOp.cu"


Differential Revision: [D25502940](https://our.internmc.facebook.com/intern/diff/D25502940)

----------
- Refactor ForeachUnaryOps.cu to better see what dtypes are supported for ops.

[ghstack-poisoned]

Commit:
Update on "Refactor ForeachUnaryOp.cu"


Differential Revision: [D25502940](https://our.internmc.facebook.com/intern/diff/D25502940)

----------
- Refactor ForeachUnaryOps.cu to better see what dtypes are supported for ops.

[ghstack-poisoned]

Commit:
Update on "Refactor ForeachUnaryOp.cu"


Differential Revision: [D25502940](https://our.internmc.facebook.com/intern/diff/D25502940)

----------
- Refactor ForeachUnaryOps.cu to better see what dtypes are supported for ops.

[ghstack-poisoned]

zou3519 commented:
@izdeby were Natalia's comments above addressed? https://github.com/pytorch/pytorch/pull/49248#discussion_r547021711 and   https://github.com/pytorch/pytorch/pull/49248#discussion_r547020802
